include/init.h

```diff
@@ -29,49 +30,101 @@ extern "C" {
 #define _SYS_INIT_LEVEL_SMP		4
 #endif
 
+struct device;
+
+/**
+ * @brief Static init entry structure for each device driver or services
+ *
+ * @param init init function for the init entry which will take the dev
+ * attribute as parameter. See below.
+ * @param dev pointer to a device driver instance structure. Can be NULL
+ * if the init entry is not used for a device driver but a service.
+ */
+struct init_entry {
+	int (*init)(struct device *dev);
+	struct device *dev;
+};
+
+void z_sys_init_run_level(s32_t level);
+
 /* A counter is used to avoid issues when two or more system devices
  * are declared in the same C file with the same init function.
  */
 #define Z_SYS_NAME(init_fn) _CONCAT(_CONCAT(sys_init_, init_fn), __COUNTER__)
 
 /**
- * @def SYS_INIT
+ * @def INIT_ENTRY_DEFINE
```

Comment:
We're introducing a new `INIT_` namespace for the defines. If I correctly remember various discussions we want to use namespaces sparingly. Since this macro should not be used directly we could prefix it with Z_ as in `Z_INIT_ENTRY_DEFINE`, or maybe `SYS_`?

Reply:
Good point, I guess Z_ would be the one